Add --ts-undefined-for-optionals command line option

ibc opened the pull request:
# Details

- Fixes #7656
- Added a new `--ts-undefined-for-optionals` command line option for `flatc`.
- If enabled, generated TypeScript code uses `undefined` for optional fields rather than `null`.
- Tests added.

ibc committed:
Add --ts-undefined-for-optionals command line option

# Details

- Fixes #7656
- Added a new `--ts-undefined-for-optionals` command line option for `flatc`.
- If enabled, generated TypeScript code uses `undefined` for optional fields rather than `null`.

ibc committed:
Also add TS generated test files

ibc committed:
Run `sh scripts/clang-format-git.sh`

ibc committed:
also add tests/ts/lalala-options.ts to the repo

ibc committed:
move new tests to tests/ts/optional_values dir

ibc committed:
add tests/ts/optional_values/optional_values_generated.cjs to the repo

ibc committed:
reuse existing optional_scalars.fbs and add new test

bjornharrtell reviewed (dismissed):
Some perhaps unnecessary source format changes but otherwise looks like solid work

ibc commented:
> Some perhaps unnecessary source format changes but otherwise looks like solid work

Could you detail which unnecessary source format changes you mean? I've run `sh scripts/clang-format-git.sh` as documented and on TS and Python files I've followed existing syntax.

bjornharrtell commented:
@aardappel I think it's is good to go but can't merge due to the test failure for swift which I would think is unrelated to this PR.

jmillan commented:
> @aardappel I think it's is good to go but can't merge due to the test failure for swift which I would think is unrelated to this PR.

The error is due to a HTTP download timeout. Probably fixable by re-running the GH action.

bjornharrtell commented:
@aardappel I was able to trigger re-run and it seems to work now.. but can't because of this:

<img width="282" height="71" alt="image" src="https://github.com/user-attachments/assets/10b39e65-cc61-4e3f-8a96-9c383a3e9e96" />

No idea what that means.

dbaileychess commented on src/idl_gen_ts.cpp:
```diff
@@ -214,6 +215,10 @@ class TsGenerator : public BaseGenerator {
 
+  // Value for non present optional fields. It's "null" by default and
+  // "undefined" if parser.ts_undefined_for_optionals flag is set.
+  std::string null_ = "null";
```
remove the initialization here, you already do it in the constructor

dbaileychess replied:
can you name this better?

dbaileychess commented on src/idl_gen_ts.cpp:
```diff
@@ -214,6 +215,10 @@ class TsGenerator : public BaseGenerator {
 
   std::map<std::string, NsDefinition> ns_defs_;
 
+  // Value for non present optional fields. It's "null" by default and
```
probably only the first sentence is needed. The code in the constructor documents the condition.

ibc committed:
add comma

ibc committed:
sh scripts/clang-format-git.sh

ibc committed:
remove comma

ibc committed:
sh scripts/clang-format-git.sh

ibc committed:
trying things

ibc committed:
sh scripts/clang-format-git.sh

ibc committed:
done

ibc committed:
address feedback

ibc committed:
sh scripts/clang-format-git.sh

ibc committed:
Merge branch 'master' into add-ts-undefined-for-optionals-command-line-argument

ibc commented:
> I'm trying to minimize changes while I am working on merging the last 6 months of changes back into Google. Hopefully I can finish tonight or tomorrow, can this wait?

Sure, no hurries. BTW I've addressed feedback already.
Thanks a lot.

jmillan committed:
Merge branch 'master' into add-ts-undefined-for-optionals-command-line-argument

ibc committed:
run `sh scripts/clang-format-git.sh`

ibc commented:
Hi @dbaileychess, I've merged master into this PR. Can we move on?

ibc committed:
remove uneeded `eslint-disable @typescript-eslint/no-namespace` line

ibc commented:
@dbaileychess, @aardappel friendly ping.

ibc committed:
Merge branch 'master' into add-ts-undefined-for-optionals-command-line-argument

ibc committed:
Merge branch 'master' into add-ts-undefined-for-optionals-command-line-argument

ibc commented:
> May I ask when a new NPM release is supposed to happen?

I mean... never?
